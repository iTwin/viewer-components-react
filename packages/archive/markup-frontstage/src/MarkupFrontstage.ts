/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import { I18N } from "@bentley/imodeljs-i18n";
import { UiError, getClassName } from "@bentley/ui-abstract";
import * as i18next from "i18next";

/**
 * Entry point for static initialization required by various components used in the package.
 * @public
 */
export class MarkupFrontstage {
  private static _i18n?: I18N;

  /**
   * Called by IModelApp to initiazlize MarkupFrontstage
   * @param i18n The name of internationalization service created by IModelAppp.
   */
  public static async initialize(i18n: I18N): Promise<void> {
    MarkupFrontstage._i18n = i18n;
    return MarkupFrontstage._i18n.registerNamespace(
      MarkupFrontstage.i18nNamespace
    ).readFinished;
  }
  /**
   * Unregisters the MakrupFrontstage internationalization service namespace.
   */
  public static terminate() {
    if (MarkupFrontstage._i18n) {
      MarkupFrontstage._i18n.unregisterNamespace(
        MarkupFrontstage.i18nNamespace
      );
    }
    MarkupFrontstage._i18n = undefined;
  }

  public static get i18n(): I18N {
    if (!MarkupFrontstage._i18n) {
      throw new UiError(
        MarkupFrontstage.loggerCategory(this),
        "MarkupFrontstage not initialized"
      );
    }
    return MarkupFrontstage._i18n;
  }

  /**
   * The internationazlization service namespace.
   */
  public static get i18nNamespace(): string {
    return "MarkupFrontstage";
  }

  public static get packageName(): string {
    return "markup-frontstage-react";
  }

  public static translate(
    key: string | string[],
    options?: i18next.TranslationOptions
  ): string {
    return MarkupFrontstage.i18n.translateWithNamespace(
      MarkupFrontstage.i18nNamespace,
      key,
      options
    );
  }

  public static loggerCategory(obj: any): string {
    const className = getClassName(obj);
    const category =
      MarkupFrontstage.packageName + (className ? `.${className}` : "");
    return category;
  }
}
