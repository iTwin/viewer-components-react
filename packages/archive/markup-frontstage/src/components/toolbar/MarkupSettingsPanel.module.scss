/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
@import "~@bentley/ui-core/lib/ui-core/style/themecolors.scss";

.markupContainer {
  display: flex;
  align-items: center;
  padding: 0;
  height: 100%;

  .markupContainerSetting:first-child {
    margin-left: 40px;
  }

  .markupSave {
    margin-left: auto;
    margin-right: 20px;
  }

  .markupSave,
  .markupCancel {
    min-width: 100px;
  }

  .markupCancel {
    margin-right: 40px;
  }

  .markupContainerSetting {
    display: flex;
    align-items: center;
    margin: 0 6px;

    > span {
      margin-right: 8px;
    }

    > .markupTransparencySlider {
      width: 110px;
    }
  }
}

.markupToolName {
  color: $buic-foreground-body;
}
