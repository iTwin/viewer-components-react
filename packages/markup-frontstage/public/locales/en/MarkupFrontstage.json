{
  "contentLayoutDefDescription": "Markup view covering the entire screen.",
  "markupSettings": {
    "cancel": "Cancel",
    "fill": "Fill",
    "opacity": "Opacity",
    "attach": "Attach",
    "size": "Size",
    "stroke": "Stroke",
    "text": "Text"
  },
  "tools": {
    "description": "Markup",
    "label": "Markup",
    "close": "Close"
  }
}
