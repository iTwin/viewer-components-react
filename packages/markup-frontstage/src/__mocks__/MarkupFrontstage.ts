/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import { I18N } from "@bentley/imodeljs-i18n";
import i18n from "../tests/mocks/i18n";

export class MarkupFrontstage {
  private static _i18n?: I18N;

  public static async initialize(i18n: I18N): Promise<void> {
    MarkupFrontstage._i18n = i18n;
  }

  public static terminate() {
    MarkupFrontstage._i18n = undefined;
  }

  public static get i18n(): I18N {
    if (!MarkupFrontstage._i18n) {
      throw new Error("IModelJs I18N not initialized");
    }
    return MarkupFrontstage._i18n;
  }

  public static get i18nNamespace(): string {
    return "MakrupFrontstage";
  }

  public static get packageName(): string {
    return "markup-frontstage-react";
  }

  public static translate(
    key: string | string[],
    options?: i18n.TranslationOptions
  ): string {
    return MarkupFrontstage.i18n.translateWithNamespace(
      MarkupFrontstage.i18nNamespace,
      key,
      options
    );
  }

  public static loggerCategory(obj: any): string {
    return obj.name;
  }
}
