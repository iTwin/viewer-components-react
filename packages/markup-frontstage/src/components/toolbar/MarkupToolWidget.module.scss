/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
.markupVerticalToolbar {
  margin-top: 11px;
}
