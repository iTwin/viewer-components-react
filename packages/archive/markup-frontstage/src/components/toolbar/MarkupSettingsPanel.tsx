/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import { ColorDef } from "@bentley/imodeljs-common";
import { MarkupApp } from "@bentley/imodeljs-markup";
import {
  AlphaSlider,
  ColorPickerButton,
  WeightPickerButton,
} from "@bentley/ui-components";
import { Point } from "@bentley/ui-core";
import { FrontstageManager, PositionPopup } from "@bentley/ui-framework";
import React, { useEffect, useState } from "react";
import { Line, Marker, SVG } from "@svgdotjs/svg.js";

import styles from "./MarkupSettingsPanel.module.scss";
import { MarkupFrontstage } from "../../MarkupFrontstage";
import { MarkupFrontstageConstants } from "../../util/MarkupTypes";

/**
 * Markup settings panel props.
 */
interface MarkupSettingsPanelProps {
  point: Point;
}

/**
 * Markup setting properties
 */
interface MarkupSettingProps {
  [key: string]: number | string;
}

/**
 * Markup data color with fill and stroke
 */
interface MarkupDataColor {
  fill: ColorDef;
  stroke: ColorDef;
}

/**
 * Line markup with marker
 */
interface LineWithMarker {
  [type: string]: Line;
}

/**
 * SVG props including SVG type and SVG CSS
 */
enum SvgProp {
  DATA_COLOR = "data-Color",
  FILL = "fill",
  FILL_OPACITY = "fill-opacity",
  LINE = "line",
  MARKER_END = "marker-end",
  MARKER_START = "marker-start",
  OPACITY = "opacity",
  STROKE = "stroke",
  STROKE_OPACITY = "stroke-opacity",
  STROKE_WIDTH = "stroke-width",
  TEXT = "text",
}

const MarkupSettingsPanel = (props: MarkupSettingsPanelProps) => {
  const { point } = props;

  const [fillColor, setFillColor] = useState<ColorDef>(
    ColorDef.create(MarkupApp.props.active.element.fill)
  );
  const [strokeColor, setStrokeColor] = useState<ColorDef>(
    ColorDef.create(MarkupApp.props.active.element.stroke)
  );
  const [textColor, setTextColor] = useState<ColorDef>(
    ColorDef.create(MarkupApp.props.active.text.fill)
  );
  const [activeWeight, setActiveWeight] = useState<number>(
    MarkupApp.props.active.element[SvgProp.STROKE_WIDTH]
  );
  const [transparency, setTransparency] = useState<number>(
    MarkupApp.props.active.element[SvgProp.STROKE_OPACITY]
  );

  // localized tool labels.
  const fill = MarkupFrontstage.translate("markupSettings.fill");
  const stroke = MarkupFrontstage.translate("markupSettings.stroke");
  const text = MarkupFrontstage.translate("markupSettings.text");
  const width = MarkupFrontstage.translate("markupSettings.size");
  const opacity = MarkupFrontstage.translate("markupSettings.opacity");

  /**
   * Creates arrow marker, ditto from iModelJs RedlineTool
   * @param color name of the color to be used to create arrow marker
   */
  const createArrowMarker = (
    color: string,
    length: number,
    width: number,
    alphaValue: number
  ): Marker => {
    // NOTE: Flashing doesn't currently affect markers, need support for "context-stroke" and "context-fill".
    // For now encode color in name...
    // Following marker Id is okay to be same, because it is used as a symbol which can be reused, however the whole
    // function will need to be updated once SVG 2.0 is available where we don't have to struggle for markers.
    const arrowMarkerId = "ArrowMarker" + length + "x" + width + "-" + color;
    let marker = SVG("#" + arrowMarkerId) as Marker;
    if (null === marker) {
      if (MarkupApp.markup) {
        marker = MarkupApp.markup
          ?.svgMarkup!.marker(length, width)
          .id(arrowMarkerId);
        marker.polygon([0, 0, length, width * 0.5, 0, width]);
        marker.attr("orient", "auto-start-reverse");
        marker.attr("overflow", "visible"); // Don't clip the stroke that is being applied to allow the specified start/end to be used directly while hiding the arrow tail fully under the arrow head...
        marker.attr("refX", length);
        marker.css({
          stroke: color,
          fill: color,
          "stroke-opacity": alphaValue,
          "fill-opacity": alphaValue,
        });
      }
    }
    return marker;
  };

  /**
   * Changes the active SVG element's property.
   * @param propName name of the property to be changed e.g. stroke, fill etc.
   * @param value name of the value to be used to change the property.
   * @param settingLabel name of the localized setting label of the tool button.
   */
  const changeActiveElementSettings = (
    propName: string,
    value: any,
    settingLabel: string
  ) => {
    // There's always single element in the selected elements set unless it's grouped.
    const selectedElements = MarkupApp.markup?.selected?.elements
      ? Array.from(MarkupApp.markup?.selected?.elements)
      : [];
    const firstSelectedElement =
      selectedElements.length > 0 ? selectedElements[0] : undefined;
    // To avoid the confusion between fill, stroke for other element with Text fill
    const isText = firstSelectedElement?.type === SvgProp.TEXT;
    const isBoxedText =
      firstSelectedElement?.type === "g" &&
      firstSelectedElement?.node?.classList.length > 0 &&
      firstSelectedElement?.node?.classList[0] === MarkupApp.boxedTextClass;
    const lineWithMarker: LineWithMarker = {};
    if (firstSelectedElement?.type === SvgProp.LINE) {
      if (
        firstSelectedElement?.node?.attributes.getNamedItem(
          SvgProp.MARKER_START
        )
      ) {
        lineWithMarker[SvgProp.MARKER_START] = firstSelectedElement as Line;
      }
      if (
        firstSelectedElement?.node?.attributes.getNamedItem(SvgProp.MARKER_END)
      ) {
        lineWithMarker[SvgProp.MARKER_END] = firstSelectedElement as Line;
      }
    }
    const markupProp = prepareMarkupProps(
      propName,
      value,
      settingLabel,
      lineWithMarker,
      isText
    );
    if (markupProp) {
      if (isBoxedText && firstSelectedElement?.children().length === 2) {
        const gRect = firstSelectedElement?.get(0);
        const gText = firstSelectedElement?.get(1);
        if (settingLabel === text && gText) {
          gText.css(markupProp);
        } else {
          if (gRect) {
            gRect.css(markupProp);
          }
          if (gText && settingLabel === opacity) {
            gText.css(markupProp);
          }
        }
      } else {
        firstSelectedElement?.css(markupProp);
      }
    }
  };

  /**
   * Helper to change active element settings.
   * @param propName name of the property to be changed e.g. stroke, fill etc.
   * @param value name of the value to be used to change the property.
   * @param settingLabel name of the localized setting label of the tool button.
   * @param lineWithMarker name of the line with marker map for arrows and distance.
   * @param isText name of the flag whether or not the element is of type text.
   */
  const prepareMarkupProps = (
    propName: string,
    value: any,
    settingLabel: string,
    lineWithMarker: LineWithMarker,
    isText?: boolean
  ): MarkupSettingProps | undefined => {
    if (isText && !(settingLabel === text || settingLabel === opacity)) {
      return undefined;
    }
    switch (propName) {
      case SvgProp.STROKE:
        MarkupApp.props.active.element[SvgProp.STROKE] = value as string;
        break;
      case SvgProp.STROKE_WIDTH:
        MarkupApp.props.active.element[SvgProp.STROKE_WIDTH] = value as number;
        break;
      case SvgProp.FILL:
        if (settingLabel === text) {
          MarkupApp.props.active.text[SvgProp.FILL] = value as string;
        } else {
          MarkupApp.props.active.element[SvgProp.FILL] = value as string;
        }
        break;
      case SvgProp.OPACITY:
        MarkupApp.props.active.element[SvgProp.FILL_OPACITY] = value as number;
        MarkupApp.props.active.element[
          SvgProp.STROKE_OPACITY
        ] = value as number;
        break;
      default:
        break;
    }
    const markupProp: MarkupSettingProps = {};
    // opacity is changed using slider so to leave hilite, emptying for non-opacity props only.
    // In case of line, fill chagnes marker only not the line itself, stroke width changes fill color of marker as bonus
    // Hence to avoid those conflicting result, only allowing marker color to be changed via stroke as follow.
    if (
      (lineWithMarker[SvgProp.MARKER_START] ||
        lineWithMarker[SvgProp.MARKER_END]) &&
      (settingLabel === stroke || settingLabel === opacity)
    ) {
      const arrowProps = MarkupApp.props.active.arrow;
      const lineDataColor = parseDataColor(
        lineWithMarker[SvgProp.MARKER_END].attr(SvgProp.DATA_COLOR)
      );
      const strokeVal =
        settingLabel === stroke ? value : lineDataColor?.stroke.name;
      const opacityVal = settingLabel === opacity ? value : 0.75;
      const changedArrow = createArrowMarker(
        strokeVal,
        arrowProps.length,
        arrowProps.width,
        opacityVal
      );
      lineWithMarker[SvgProp.MARKER_END].marker("end", changedArrow);
      lineWithMarker[SvgProp.MARKER_START]?.marker("start", changedArrow);
      if (settingLabel === opacity) {
        markupProp[SvgProp.FILL_OPACITY] = value;
        markupProp[SvgProp.STROKE_OPACITY] = value;
        markupProp[propName] = value;
        return markupProp;
      }
    }
    if (propName === SvgProp.OPACITY) {
      markupProp[SvgProp.FILL_OPACITY] = value;
      markupProp[SvgProp.STROKE_OPACITY] = value;
    } else {
      MarkupApp.markup?.selected?.emptyAll();
      markupProp[propName] = value;
    }
    return markupProp;
  };

  /**
   * Helper to parse data color.
   * @param dataColor name of the data color string with fill and stroke
   */
  const parseDataColor = (dataColor: string): MarkupDataColor | undefined => {
    if (
      !dataColor.includes(SvgProp.FILL) &&
      !dataColor.includes(SvgProp.STROKE)
    ) {
      return undefined;
    }
    const dataColorPayload = JSON.parse(dataColor);
    return {
      fill: ColorDef.create(dataColorPayload.fill),
      stroke: ColorDef.create(dataColorPayload.stroke),
    };
  };

  useEffect(() => {
    // There's always single element in selected element unless its grouped.
    const selectedElements = MarkupApp.markup?.selected?.elements
      ? Array.from(MarkupApp.markup?.selected?.elements)
      : [];
    const firstSelectedElement =
      selectedElements?.length > 0 ? selectedElements[0] : undefined;
    if (firstSelectedElement?.inSelection) {
      // adding length check as well as classList check to confirm boxedText.
      const isBoxedTextGType =
        firstSelectedElement?.type === "g" &&
        firstSelectedElement?.node.classList.length > 0 &&
        firstSelectedElement?.node.classList[0] === MarkupApp.boxedTextClass &&
        firstSelectedElement?.children().length === 2;
      // using strokeOpacity instead of fill because line doesn't have fillOpacity.
      const alphaVal = isBoxedTextGType
        ? firstSelectedElement?.get(0)?.node?.style.strokeOpacity
        : firstSelectedElement?.type === "line" &&
          firstSelectedElement?.node.style.opacity
        ? firstSelectedElement?.node.style.opacity
        : firstSelectedElement?.node.style.strokeOpacity;
      const strokeWidth = isBoxedTextGType
        ? firstSelectedElement?.get(0)?.node?.style.strokeWidth
        : firstSelectedElement?.node.style.strokeWidth;
      // In case if opacity and stroke are not present then setting it to default 0.75 and 3 respectively.
      // Markup app has transparency setup for both fill and stroke, by default fill is 0.2 and stroke is 0.8
      // In case if user wants it to have similar effect then we'll need to have two alpha slider
      setTransparency(alphaVal ? parseFloat(alphaVal) : 0.75);
      setActiveWeight(strokeWidth ? parseInt(strokeWidth) : 3);
      const dataColorAttr = isBoxedTextGType
        ? firstSelectedElement?.get(0)?.attr()[SvgProp.DATA_COLOR]
        : firstSelectedElement?.attr(SvgProp.DATA_COLOR);
      if (dataColorAttr) {
        const dataColor = parseDataColor(dataColorAttr);
        if (dataColor) {
          if (firstSelectedElement?.type === SvgProp.TEXT) {
            setTextColor(dataColor.fill);
            // setting default fill color to blue and stroke to red other wise it gets reset to black.
            setFillColor(ColorDef.create("rgb(0, 0, 255)"));
            setStrokeColor(ColorDef.create("rgb(255, 0, 0)"));
          } else {
            setFillColor(dataColor.fill);
            setStrokeColor(dataColor.stroke);
          }
        }
      }
      if (isBoxedTextGType) {
        const textColorAttr = firstSelectedElement?.get(1)?.attr()[
          SvgProp.DATA_COLOR
        ];
        if (textColorAttr) {
          const textColor = parseDataColor(textColorAttr);
          if (textColor) {
            setTextColor(textColor.fill);
          }
        }
      }
    }
  }, [
    FrontstageManager.activeFrontstageDef?.findWidgetDef(
      MarkupFrontstageConstants.WIDGET_ID
    )?.state,
  ]);

  return (
    <div data-testid={"markup-settings-tool"}>
      <PositionPopup point={point}>
        <div className={styles.markupContainer}>
          <div
            className={styles.markupContainerSetting}
            data-testid={"stroke-color-picker-button"}
          >
            <span className={styles.markupToolName}>{stroke}</span>
            <ColorPickerButton
              initialColor={strokeColor}
              onColorPick={(color: ColorDef) => {
                if (color.name) {
                  changeActiveElementSettings(
                    SvgProp.STROKE,
                    color.name,
                    stroke
                  );
                }
                setStrokeColor(color);
              }}
            />
          </div>
          <div
            className={styles.markupContainerSetting}
            data-testid={"fill-color-picker-button"}
          >
            <span className={styles.markupToolName}>{fill}</span>
            <ColorPickerButton
              initialColor={fillColor}
              onColorPick={(color: ColorDef) => {
                if (color.name) {
                  changeActiveElementSettings(SvgProp.FILL, color.name, fill);
                }
                setFillColor(color);
              }}
            />
          </div>
          <div
            className={styles.markupContainerSetting}
            data-testid={"stroke-weight-picker-button"}
          >
            <span className={styles.markupToolName}>{width}</span>
            <WeightPickerButton
              activeWeight={activeWeight}
              onLineWeightPick={(weight: number) => {
                changeActiveElementSettings(
                  SvgProp.STROKE_WIDTH,
                  weight,
                  width
                );
                setActiveWeight(weight);
              }}
            />
          </div>
          <div
            className={styles.markupContainerSetting}
            data-testid={"alpha-slider"}
          >
            <span className={styles.markupToolName}>{opacity}</span>
            <AlphaSlider
              className={styles.markupTransparencySlider}
              alpha={transparency}
              isHorizontal={true}
              onAlphaChange={(alpha: number) => {
                changeActiveElementSettings(SvgProp.OPACITY, alpha, opacity);
                setTransparency(alpha);
              }}
            />
          </div>
          <div
            className={styles.markupContainerSetting}
            data-testid={"text-color-picker-button"}
          >
            <span className={styles.markupToolName}>{text}</span>
            <ColorPickerButton
              initialColor={textColor}
              onColorPick={(color: ColorDef) => {
                if (color.name) {
                  changeActiveElementSettings(SvgProp.FILL, color.name, text);
                }
                setTextColor(color);
              }}
            />
          </div>
        </div>
      </PositionPopup>
    </div>
  );
};

export default MarkupSettingsPanel;
