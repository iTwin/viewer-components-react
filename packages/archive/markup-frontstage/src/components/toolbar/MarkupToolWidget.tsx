/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import {
  ActionButton,
  CommonToolbarItem,
  ToolbarItemUtilities,
  ToolbarOrientation,
  ToolbarUsage,
} from "@bentley/ui-abstract";
import {
  MarkupTools,
  ToolbarComposer,
  ToolbarHelper,
  ToolItemDef,
  ToolWidgetComposer,
} from "@bentley/ui-framework";
import React from "react";

import styles from "./MarkupToolWidget.module.scss";
import { MarkupFrontstage } from "../../MarkupFrontstage";
import { MarkupFrontstageConstants } from "../../util/MarkupTypes";
/**
 * Markup tool widget props
 */
interface MarkupToolWidgetProps {
  closeMarkupFrontstageAsync: () => void;
  isEditable: boolean;
}

export class MarkupToolWidget extends React.Component<MarkupToolWidgetProps> {
  private _verticalItems: CommonToolbarItem[] = [
    ToolbarHelper.createToolbarItemFromItemDef(10, MarkupTools.selectToolDef),
    ToolbarItemUtilities.createGroupButton(
      MarkupFrontstageConstants.DRAWING_TOOLS,
      20,
      "icon-2d",
      MarkupFrontstage.translate("tools.label"),
      [
        ToolbarHelper.createToolbarItemFromItemDef(
          10,
          MarkupTools.lineToolDef,
          { id: "line-tool" }
        ) as ActionButton,
        ToolbarHelper.createToolbarItemFromItemDef(
          20,
          MarkupTools.rectangleToolDef,
          { id: "rectangle-tool" }
        ) as ActionButton,
        ToolbarHelper.createToolbarItemFromItemDef(
          30,
          MarkupTools.polygonToolDef
        ) as ActionButton,
        ToolbarHelper.createToolbarItemFromItemDef(
          40,
          MarkupTools.cloudToolDef
        ) as ActionButton,
        ToolbarHelper.createToolbarItemFromItemDef(
          50,
          MarkupTools.ellipseToolDef
        ) as ActionButton,
        ToolbarHelper.createToolbarItemFromItemDef(
          60,
          MarkupTools.arrowToolDef
        ) as ActionButton,
        ToolbarHelper.createToolbarItemFromItemDef(
          70,
          MarkupTools.sketchToolDef
        ) as ActionButton,
      ]
    ),
    ToolbarHelper.createToolbarItemFromItemDef(
      30,
      MarkupTools.placeTextToolDef
    ),
    ToolbarHelper.createToolbarItemFromItemDef(40, MarkupTools.distanceToolDef),
  ];

  /** Close Markup Frontstage tool def */
  private _closeMarkupToolDef = new ToolItemDef({
    toolId: "Markup.Close",
    iconSpec: "icon-close",
    label: () => MarkupFrontstage.translate("tools.close"),
    tooltip: () => MarkupFrontstage.translate("tools.close"),
    execute: async () => {
      this.props.closeMarkupFrontstageAsync();
    },
  });

  public render() {
    // adding following style because row-gap is added
    // in MainFrontstage.overrides.scss for identifier nz-vertical-toolbar
    const { isEditable } = this.props;
    return (
      <div data-testid="markup-tool-widget-container">
        <ToolWidgetComposer
          verticalToolbar={
            <ToolbarComposer
              items={
                isEditable
                  ? this._verticalItems
                  : [
                      ToolbarHelper.createToolbarItemFromItemDef(
                        10,
                        this._closeMarkupToolDef
                      ),
                    ]
              }
              usage={ToolbarUsage.ContentManipulation}
              orientation={ToolbarOrientation.Vertical}
            />
          }
          className={styles.markupVerticalToolbar}
          data-testid={"markup-tool-widget"}
        />
      </div>
    );
  }
}
