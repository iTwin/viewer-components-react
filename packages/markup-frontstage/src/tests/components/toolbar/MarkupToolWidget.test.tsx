/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import { I18N } from "@bentley/imodeljs-i18n";
import { UiFramework } from "@bentley/ui-framework";
import "@testing-library/jest-dom/extend-expect";
import { render } from "@testing-library/react";
import React from "react";
import configureMockStore from "redux-mock-store";
import { MarkupToolWidget } from "../../../components/toolbar/MarkupToolWidget";
import { MarkupFrontstage } from "../../../MarkupFrontstage";

jest.mock("@bentley/imodeljs-frontend");
jest.mock("@bentley/imodeljs-i18n", () => ({
  I18N: jest.fn().mockImplementation(() => {
    return {
      translateWithNamespace: () => (props: { defaultProps: any }) => {
        props.defaultProps = {
          ...props.defaultProps,
          translate: () => "",
        };
        return props;
      },
      registerNamespace: () => {
        return jest.fn();
      },
      unregisterNamespace: () => {
        return jest.fn();
      },
    };
  }),
}));
jest.mock("@bentley/ui-components");
jest.mock("../../../MarkupFrontstage");

const mockStore = configureMockStore([]);

describe("Markup Tool Widget testing", () => {
  let store: any;

  beforeEach(async () => {
    store = mockStore({
      sampleAppState: { placeHolder: false },
      testDifferentFrameworkKey: {
        configurableUiState: {
          snapMode: 1,
          toolPrompt: "test tool prompt",
          theme: "test-theme",
          widgetOpacity: 100,
        },
        sessionState: {
          numItemsSelected: 0,
          availableSelectionScopes: [
            { id: "testId", label: "testPresentationSelectionScope" },
          ],
          activeSelectionScope: "testScope",
          iModelId: "testIModelId",
        },
      },
    });
    await UiFramework.initialize(store, new I18N());
  });

  afterEach(() => {
    UiFramework.terminate();
  });

  it("should render", () => {
    //arrange
    MarkupFrontstage.initialize(new I18N());
    const { queryByTestId } = render(
      <MarkupToolWidget
        isEditable={false}
        closeMarkupFrontstageAsync={jest.fn()}
      />
    );
    //act
    const toolWidget = queryByTestId("markup-tool-widget-container");

    //expect
    expect(toolWidget).toBeInTheDocument();
    expect(toolWidget?.firstChild).toHaveClass("markupVerticalToolbar");
  });
});
