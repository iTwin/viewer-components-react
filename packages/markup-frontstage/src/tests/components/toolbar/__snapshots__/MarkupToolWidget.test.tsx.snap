// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markup Tool Widget testing should render 1`] = `
<div
  data-testid="markup-tool-widget-container"
>
  <div
    class="nz-tools-widget verticalToolbar"
  >
    <div
      class="nz-vertical-tool-area"
    >
      <div
        class="nz-app-button"
      />
      <div
        class="nz-vertical-toolbar-container"
      >
        <div
          class="nz-toolbar-toolbar nz-direction-right nz-vertical nz-panel-alignment-start"
        >
          <div
            class="nz-expanded nz-panels"
          >
            <div>
              <div
                class="nz-panel"
              />
            </div>
          </div>
          <div
            class="nz-toolbar-items nz-vertical nz-items"
          >
            <button
              class="nz-toolbar-item-item"
            >
              <div
                class="nz-icon"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="nz-horizontal-toolbar-container"
    />
  </div>
</div>
`;
